[MRG+1] DOC: Add missing cmaps to perception doc (fix for #8073)

afvincent opened the pull request:
Fixes #8073 .

Add the missing `tab10` and `tab20*` colormaps to the relevant [web page](http://matplotlib.org/devdocs/users/colormaps.html), that are currently missing as noticed by #8073 . I also added the `Wistia` colormap that also appears to be missing from this documentation.

I did a minor clean up of the `grayscale.py` script (removing some outdated comments for example). Furthermore, the last commit is an overhaul of the `lightness.py` script, aiming at making it a bit more resilient to adding or deleting one colormap map without having to tweak a lot of "magical" constant values determined by hand. I manly tried to factorize some parts when it seemed possible to do some.

The documentation is still building well :). Here is a self-contained archive version of the `colormaps.html` web page, with the updated plots: [Choosing_Colormaps_doc.zip](https://github.com/matplotlib/matplotlib/files/801966/Choosing_Colormaps_doc.zip). For example, the figures with the Vega-related and Wistia colormaps:
![grayscale_figure_3_sequential_2_colormaps](https://cloud.githubusercontent.com/assets/17270724/23339960/5f40de24-fc2d-11e6-8d40-1ceb02d85411.png)
![grayscale_figure_5_qualitative_colormaps](https://cloud.githubusercontent.com/assets/17270724/23339962/5f45b962-fc2d-11e6-9bd2-c0430a6829ce.png)
![lightness_figure_3_sequential_2_colormaps](https://cloud.githubusercontent.com/assets/17270724/23339961/5f4533e8-fc2d-11e6-9afb-93950b12ceef.png)
![lightness_figure_5_qualitative_colormaps](https://cloud.githubusercontent.com/assets/17270724/23339963/5f4db7d4-fc2d-11e6-8280-b21c26c2880c.png)

*If this was to be merged, please backport it to `2.0.0-doc`* (See @tacaswell's comment in #8073). I am still not totally friend with git and did not manage to directly target the `2.0.0-doc` branch 🐑 ...

**Edit:** added a keyword to automatically close the related issue.

afvincent commented:
I am not sure that I read the `codecov` logs correctly, but as far as I understand them, the failure is related to changes done by commits that are different from the ones of this PR.

phobson commented on doc/users/plotting/colormaps/colormaps.py:
```diff
@@ -7,14 +7,15 @@
-                             'pink', 'spring', 'summer', 'winter']),
+         ('Sequential (2)', ['afmhot', 'autumn', 'bone', 'cool', 'copper',
+                             'gist_heat', 'gray', 'hot', 'pink', 'spring',
+                             'Wistia', 'summer', 'winter']),
```
Taking a step back, I think we have three subgroups of sequential colormaps:

  1. single color
  2. blended colors
  3. temperature/seasons

I guess I'd really like to see plots for each of these groups. 

Opinion: And within each group, we should sort/group as logically as possible, as opposed to alphabetically

afvincent replied:
See 3c29961: I tried to group the colormap in a more logical or hue-base fashion, rather than per alphabetical order.

phobson commented on doc/users/plotting/colormaps/lightness.py:
```diff
@@ -1,120 +1,100 @@
 
-# indices to step through colormap
+# Number of colormap par subplot for particular cmap categories
+_DSUBS = {'Perceptually Uniform Sequential': 4, 'Sequential': 6,
```
It'd be nice if we could compute the values in someway. I that would require the colors module to be reorganzed in some way. And that's definitely out of the scope of this PR.

afvincent replied:
The new order (and the addition of some previously missing colormaps) in 3c29961 allow a more uniform set of values for `_DSUBS` and `_DC`. Basically only two different values are sufficient to clean reference plots. 

I have kept an explicit dictionary with all values inside, in case one would want to further tweak these values. One could also use these dictionary to just describe the non-standard cases and rely on `.get(key, default)` for the others.

phobson commented on doc/users/plotting/colormaps/lightness.py:
```diff
@@ -1,120 +1,100 @@
 mpl.rcParams.update({'font.size': 12})
 
-# indices to step through colormap
+# Number of colormap par subplot for particular cmap categories
```
"par" -> "per" ?

afvincent replied:
Yep (French went through the cracks of my mind…).

afvincent replied:
Fixed!

afvincent commented:
@phobson Thank you for the feedback :). I will try to see what is feasible this weekend.

NelleV commented:
@afvincent don't hesitate to ping us when it is done. This PR disappeared from the frontpage.

afvincent commented:
@NelleV Ok and don't worry, it is still in my to-do list. I was just (and am still) trying to finish the other stuff I am working on in parallel, like the eventplot stuff (with your comments waiting to be addressed for example ;) ).

afvincent commented:
Removing the `need_review` flag for the moment because I still have to see what can be done about @phobson's suggestions, and some other on-going PRs that I have are likely to require more time than I expected in the next few days. I will ping when this PR will be ready again for review.

**Note to self:** I think *Wistia* is missing from the colormaps that I demonstrated in the gallery example.

afvincent commented:
Rebased and added some missing colormaps (`binary`, `gist_gray` and `gist_yarg`). Also tried to group the colormaps in a more logical fashion.

I do not really understand why, but I modified `doc/examples/color/colormaps_reference.rst` (which more or less looks like a copy of `doc/mpl_examples/color/colormaps_reference.py`) and it appears to not be tracked by git. Is it expected?

Building the docs results in:
[Choosing Colormaps — Matplotlib 2.0.0.post3789.dev0+g0c9a2805f documentation.zip](https://github.com/matplotlib/matplotlib/files/869968/Choosing.Colormaps.Matplotlib.2.0.0.post3789.dev0.g0c9a2805f.documentation.zip).
The plots for people allergic to zip:
[lightness_00.pdf](https://github.com/matplotlib/matplotlib/files/869970/lightness_00.pdf)
[lightness_01.pdf](https://github.com/matplotlib/matplotlib/files/869971/lightness_01.pdf)
[lightness_02.pdf](https://github.com/matplotlib/matplotlib/files/869972/lightness_02.pdf)
[lightness_03.pdf](https://github.com/matplotlib/matplotlib/files/869974/lightness_03.pdf)
[lightness_04.pdf](https://github.com/matplotlib/matplotlib/files/869973/lightness_04.pdf)
[lightness_05.pdf](https://github.com/matplotlib/matplotlib/files/869975/lightness_05.pdf)
and
[grayscale_01_00.pdf](https://github.com/matplotlib/matplotlib/files/869976/grayscale_01_00.pdf)
[grayscale_01_01.pdf](https://github.com/matplotlib/matplotlib/files/869977/grayscale_01_01.pdf)
[grayscale_01_02.pdf](https://github.com/matplotlib/matplotlib/files/869978/grayscale_01_02.pdf)
[grayscale_01_03.pdf](https://github.com/matplotlib/matplotlib/files/869981/grayscale_01_03.pdf)
[grayscale_01_04.pdf](https://github.com/matplotlib/matplotlib/files/869980/grayscale_01_04.pdf)
[grayscale_01_05.pdf](https://github.com/matplotlib/matplotlib/files/869979/grayscale_01_05.pdf)

Ping to @NelleV and @phobson

NelleV commented:
the doc/examples folder is generated from the examples. The fact that it is not tracked is to be expected.

NelleV commented:
Once the sphinx-gallery PR is merged, those python script should really be moved to our examples folder.
I'll do a full review after coffee :D

QuLogic commented on doc/users/plotting/colormaps/lightness.py:
```diff
@@ -1,120 +1,100 @@
+          'Sequential (2)': 6, 'Diverging': 6, 'Qualitative': 4,
+          'Miscellaneous': 6}
+
+# Spacing between the colormaps of a subplot
```
I think it makes more sense to change the figure size like it used to be done. Now you get a different sized Axes and different number of tweaks for each figure.

afvincent replied:
I do not understand to which situation you are refering to: looking at [blame](https://github.com/afvincent/matplotlib/blame/a769e5b3afe317f8626738f611c33bfb29f2b6b8/doc/users/plotting/colormaps/lightness.py#L23) on GH, the `dc` and `dsub` strategy is used since at least 3 years. I tried to not modify the spirit of the plots, but rather to refactor what could be, and to make the magic values a bit more uniform across all the different cases.

QuLogic replied:
Oh, this comment probably shouldn't be here but on a `subplots` line. In any case, I missed the fact that you kept the changing figure size, so it should be the same as before.

But if the figure size changes based on the number of plots, then I don't understand why 00 and 05 have 3 y-ticks, but the others have 5.

afvincent replied:
I did not even notice the different amount of ticks ^^... If I had to guess, I would say that it may be an artifact from `plt.tight_layout()` that gives more or less place to the subplots depending on length of the x-tick labels.

QuLogic replied:
That might be it. It would be nice if we didn't have to use `tight_layout`, but I guess then we'd need to know the size of the colormap labels first.

afvincent commented:
Please do not merge yet, I saw a minor typo (l.37 on `lightness.py`). I'll try to fix it as soon as I get back to my laptop.

dstansby commented:
Backported to `2.0.x` via. https://github.com/matplotlib/matplotlib/commit/8be411f466f5ad219951ce78b3e513187c9d217d
